iceoryx_hoofs/test/moduletests/test_cxx_function.cpp

```diff
@@ -0,0 +1,630 @@
+    auto nonStorable = []() {};
+    using NonStorable = decltype(nonStorable);
+    constexpr auto BYTES = test_function::required_storage_size<NonStorable>();
+    constexpr auto RESULT = iox::cxx::function<signature, BYTES>::is_storable<NonStorable>();
+    EXPECT_FALSE(RESULT);
+}
+
+
+TEST_F(function_test, callWithCopyConstructibleArgument)
+{
+    iox::cxx::function<int32_t(Arg), 1024> sut(freeFunctionWithCopyableArg);
+    std::function<int32_t(Arg)> func(freeFunctionWithCopyableArg);
+    Arg::resetCounts();
+
+    Arg arg(73);
+
+    auto result = sut(arg);
+
+    EXPECT_EQ(result, freeFunctionWithCopyableArg(arg));
+    EXPECT_EQ(result, func(arg));
+    // note that by using the numCopies counter we can observe that the std::function call also performs 2 copies of arg
+    // in this case
+}
+
+TEST_F(function_test, callWithVoidSignatureWorks)
+{
+    const int32_t initial = 73;
+    int value = initial;
+    auto lambda = [&]() { ++value; };
+    iox::cxx::function<void(void), 128> sut(lambda);
+
+    ASSERT_TRUE(sut.operator bool());
+    sut();
+
+    EXPECT_EQ(value, initial + 1);
+}
+
+TEST_F(function_test, callWithReferenceArgumentsWorks)
+{
+    const int32_t initial = 73;
+    Arg arg(initial);
+
+    auto lambda = [](Arg& a) { ++a.value; };
+    iox::cxx::function<void(Arg&), 128> sut(lambda);
+
+    ASSERT_TRUE(sut.operator bool());
+    sut(arg);
+
+    EXPECT_EQ(arg.value, initial + 1);
+}
+
+TEST_F(function_test, callWithConstReferenceArgumentsWorks)
+{
+    const int32_t initial = 73;
+    Arg arg(initial);
+
+    auto lambda = [](const Arg& a) { return a.value + 1; };
+    iox::cxx::function<int32_t(const Arg&), 128> sut(lambda);
+
+    ASSERT_TRUE(sut.operator bool());
+    auto result = sut(arg);
+
+    EXPECT_EQ(result, initial + 1);
+}
+
+TEST_F(function_test, callWithValueArgumentsWorks)
+{
+    const int32_t initial = 73;
+    Arg arg(initial);
+
+    auto lambda = [](Arg a) { return a.value + 1; };
+    iox::cxx::function<int32_t(Arg&), 128> sut(lambda);
+
+    ASSERT_TRUE(sut.operator bool());
+    auto result = sut(arg);
+
+    EXPECT_EQ(result, initial + 1);
+}
+
+TEST_F(function_test, callWithRValueReferenceArgumentsWorks)
+{
+    const int32_t initial = 73;
+    Arg arg(initial);
+
+    auto lambda = [](Arg&& a) { return a.value + 1; };
+    iox::cxx::function<int32_t(Arg &&), 128> sut(lambda);
+
+    ASSERT_TRUE(sut.operator bool());
+    auto result = sut(std::move(arg));
+
+    EXPECT_EQ(result, initial + 1);
+}
+
+TEST_F(function_test, callWithMixedArgumentsWorks)
+{
+    Arg arg1(1);
+    Arg arg2(2);
+    Arg arg3(3);
+    Arg arg4(4);
+
+    constexpr int32_t sum = 10;
+
+    auto lambda = [](Arg& a1, const Arg& a2, Arg&& a3, Arg a4) { return a1.value + a2.value + a3.value + a4.value; };
+    iox::cxx::function<int32_t(Arg&, const Arg&, Arg&&, Arg), 128> sut(lambda);
+
+    ASSERT_TRUE(sut.operator bool());
+    auto result = sut(arg1, arg2, std::move(arg3), arg4);
+
+    EXPECT_EQ(result, sum);
+}
+
```

Comment:
You could add some tests to see if SFINAE is working properly, like [those tests](https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_hoofs/test/moduletests/test_cxx_function_ref.cpp#L304-L320).

Reply:
Why? SFINAE is an implementation detail (compile time technique to switch between implementations), not a requirement of the behavior of the object.

Besides, I do not quite see the point of the tests you linked. They test whether I can pass some callable to a function which takes a `function_ref` (of compatible signature) by value. Internally the `function_ref` is then called.
Would the signature not be compatible, it would not compile.

This is equivalent to just pass the callable to the function_ref itself and callng it, no indirection with an additional function is needed nor useful. The equivalent case mentioned is already tested for `cxx::function` by testing the ctor itself.

TLDR: the tests add nothing since the ctor is already tested so I will add no such test.

Reply:
What I meant is, to add some tests to check if the overload resolution works as expected. I remember having missed the tests in the first attempt of `function_ref` and when using it I had to narrow down possible set during the overload resolution via type traits.